File: Computational_Physics_2/1.Lectures/7.Computacion_en_parallelo_MPI/examples/examples_C/example_4_Bcast.c
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {

   int nprocs,rang;
   int n;

   MPI_Init(&argc,&argv);
   MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
   MPI_Comm_rank(MPI_COMM_WORLD, &rang);

   if (rang==0) {
     FILE *fp = fopen("data.txt", "r");
     fscanf(fp, "%d", &n);
     fclose(fp);
   }

   MPI_Bcast(&n,1,MPI_INTEGER,0,MPI_COMM_WORLD);

   printf("Process %d has received %d \n",rang,n);

   MPI_Finalize();

   return 0;
}

File: Computational_Physics_2/1.Lectures/7.Computacion_en_parallelo_MPI/examples/examples_fortran/example_4_Bcast.f90
! https://indico.in2p3.fr/event/20625/contributions/101997/attachments/67854/95270/mpi.pdf
PROGRAM bcast
  USE mpi
  IMPLICIT NONE

  INTEGER :: n,nprocs,rang
  INTEGER :: ierr = 0
  INTEGER, PARAMETER :: idat=91

  CALL MPI_INIT(ierr)
  CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nprocs, ierr)
  CALL MPI_COMM_RANK(MPI_COMM_WORLD, rang, ierr)

  IF (rang.eq.0) THEN
    OPEN(unit=idat,file='data.txt')
    READ(idat,*) n
    CLOSE(idat)
  END IF

  CALL MPI_BCAST(n,1,MPI_INT,0,MPI_COMM_WORLD,ierr)

  print *,'Process',rang,' has received ',n

  CALL MPI_FINALIZE(ierr)

END PROGRAM bcast
